albedo.js

```diff
@@ -28,55 +29,79 @@ module.exports = {
       insecureAuth: true,
     });
 
-    connection.query(options.query, (err, rows) => {
-      if (err) {
-        return callback(err);
-      }
-      if (rows.length === 0) {
-        return callback('No records for query');
-      }
+    // prune older reports
+    if (options.hasOwnProperty('removeOlderThan')) {
+      rmDir(options.location, options);
+    }
 
-      let processedRows = rows;
-      // Allow calling code to pass either a function or an array of functions
-      // with which to process each row of data
-      if (_.isArray(options.process_row)) {
-        _.each(options.process_row, (func) => {
-          processedRows = _.map(processedRows, func);
-        });
-      } else if (_.isFunction(options.process_row)) {
-        processedRows = _.map(processedRows, options.process_row);
-      }
+    // set up input, csv, and output streams
+    const fileName = `${options.name}_${moment().format('YYYY-MM-DD_HH-mm-ss')}.csv`;
+    const outputPath = `${options.location}/${fileName}`;
 
-      return json2csv(
-        {
-          data: processedRows,
-          preserveNewLinesInValues: true,
-        },
-        (err1, csv) => {
-          if (err1) {
-            return callback(err1);
-          }
+    const input = connection.query(options.query).stream({ highWaterMark: 64 });
+    const output = fs.createWriteStream(outputPath, { encoding: 'utf8' });
+    const json2csv = new CSVTransform(null, { objectMode: true });
 
-          if (options.hasOwnProperty('removeOlderThan')) {
-            rmDir(options.location, options);
-          }
-          // make the new report
-          const fileName = `${options.name}_${moment().format('YYYY-MM-DD_HH-mm-ss')}.csv`;
-          fs.writeFile(`${options.location}/${fileName}`, csv, (err2) => {
-            if (err2) {
-              return callback(err2);
-            }
-            const reportInfo = {
-              name: fileName,
-              path: `${options.location}/`,
-            };
+    // handle row transformations
+    let empty = true;
+    const rowTransform = new Transform({
+      writableObjectMode: true,
+      readableObjectMode: true,
 
-            return callback(null, reportInfo);
-          });
+      transform(chunk, enc, handler) {
+        try {
+          let row = chunk;
+          if (_.isArray(options.process_row)) {
+            _.each(options.process_row, (func) => {
+              row = func(row);
+            });
+          } else if (_.isFunction(options.process_row)) {
+            row = options.process_row(row);
+          }
+          handler(null, row);
+          empty = false;
```

Comment:
I'm not 100% sure this is correct, will this get set even though you are setting it after calling the callback? I notice there's no return on the callback, so maybe it will, but I wonder if it should go first.

Reply:
I deferred setting `empty` until after the handler is called in case subsequent handlers in the pipeline just throw errors continually. The transform's handler works async and doesn't have a return value, so that line definitely gets executed (I verified that behavior by separately testing queries that return zero and return at least one result).